// astronomy-bundle/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.timeCalc = exports.sunCalc = exports.starCalc = exports.precessionCalc = exports.observationCalc = exports.moonPhaseCalc = exports.moonCalc = exports.earthCalc = exports.distanceCalc = exports.coordinateCalc = exports.conjunctionCalc = exports.angleCalc = void 0;
var calculations_1 = require("../earth/calculations");
Object.defineProperty(exports, "earthCalc", { enumerable: true, get: function () { return calculations_1.earthCalc; } });
var calculations_2 = require("../moon/calculations");
Object.defineProperty(exports, "moonCalc", { enumerable: true, get: function () { return calculations_2.moonCalc; } });
var calculations_3 = require("../moon/calculations");
Object.defineProperty(exports, "moonPhaseCalc", { enumerable: true, get: function () { return calculations_3.moonPhaseCalc; } });
var calculations_4 = require("../sun/calculations");
Object.defineProperty(exports, "sunCalc", { enumerable: true, get: function () { return calculations_4.sunCalc; } });
var calculations_5 = require("../time/calculations");
Object.defineProperty(exports, "timeCalc", { enumerable: true, get: function () { return calculations_5.timeCalc; } });
var calculations_6 = require("../stars/calculations");
Object.defineProperty(exports, "starCalc", { enumerable: true, get: function () { return calculations_6.starCalc; } });
var calculations_7 = require("../planets/calculations");
Object.defineProperty(exports, "conjunctionCalc", { enumerable: true, get: function () { return calculations_7.conjunctionCalc; } });
var calculations_8 = require("../coordinates/calculations");
Object.defineProperty(exports, "coordinateCalc", { enumerable: true, get: function () { return calculations_8.coordinateCalc; } });
Object.defineProperty(exports, "precessionCalc", { enumerable: true, get: function () { return calculations_8.precessionCalc; } });
var angleCalc = __importStar(require("./angleCalc"));
exports.angleCalc = angleCalc;
var distanceCalc = __importStar(require("./distanceCalc"));
exports.distanceCalc = distanceCalc;
var observationCalc = __importStar(require("./observationCalc"));
exports.observationCalc = observationCalc;
